docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+
+- the relayer send headers and data to the handler to invoke
+  `checkValidityAndUpdateState` and
+  `checkMisbehaviourAndUpdateState`. It may also query
+  `queryChainConsensusState`.
+  
+- multiple relayers may talk to one handler. Some relayers might be 
+  faulty. We assume existence of at least single correct relayer.
+
+
+## Informal Problem Statement: Fork detection in IBC
+  
+### Relayer requirement: Evidence for Handler
+
+- The relayer should provide the handler with
+  "evidence" that there was a fork. 
+  
+- The relayer can read the handler's consensus state. Thus the relayer can
+  feed the handler precisely the information the handler needs to detect a
+  fork. 
+  What is this
+  information needs to be specified.
+  
+- The information depends on the verification the handler does. It
+  might be necessary to provide a bisection proof (list of
+  lightblocks) so that the handler can verify based on its local
+  lightstore a header *h* that is conflicting with a header *h'* in the
+  local lightstore, that is, *h != h'* and *h.Height = h'.Height*
+  
+### Relayer requirement: Fork detection
+
+Let's assume there is a fork at chain A. There are two ways the
+relayer can figure that out:
+
+1. as the relayer contains a light client for A, it also includes a fork
+   detector that can detect a fork.
+	  
+2. the relayer may also detect a fork by observing that the
+   handler for chain A (on chain B)
+   is on a different branch than the relayer
+
+
+- in both detection scenarios, the relayer should submit evidence to 
+  full nodes of chain A where there is a fork. As we assume a fullnode
+  has a complete list of blocks, it is sufficient to send "Bucky's
+  evidence" (https://github.com/tendermint/tendermint/issues/5083),
+  that is, 
+     - two lightblocks from different branches + 
+	 - a lightblock (perhaps just a height) from which both blocks 
+	   can be verified.
+  
+- in the scenario 2., the relayer must feed the A-handler (on chain B)
+  a proof of a fork on A so that chain B can react accordingly
+  
+### Handler requirement
+  
+- there are potentially many relayers, some correct some faulty
+
+- a handler cannot trust the information provided by the relayer, 
+  but must verify 
+  (Доверя́й, но проверя́й)
+
+- in case of a fork, we accept that the handler temporarily stores
+  headers (tagged as verified). 
+  
+- eventually, a handler should be informed
+ (`checkMisbehaviourAndUpdateState`)
+ by some relayer that it has
+  verified a header from a fork. Then the handler should do what is
+ required by IBC in this case (stop?)
+
+### Challenges in the handler requirement
+
+- handlers and relayers work on different lightstores. In principle
+  the lightstore need not intersect in any heights a priori
+
+- if a relayer  sees a header *h* it doesn't know at a handler (`queryChainConsensusState`), the
+  relayer needs to
+  verify that header. If it cannot do it locally based on downloaded
```

Comment:
so, according to the [ics 07 spec](https://github.com/cosmos/ics/tree/master/spec/ics-007-tendermint-client#consensus-state) and the [current sdk implementation](https://github.com/cosmos/cosmos-sdk/blob/master/x/ibc/07-tendermint/types/consensus_state.go#L15), the ConsensusState only contains a validator set, it doesn't actually contain the full commit or even the header (and it probably shouldn't, that's a lot of data to store on chain). So unless this were to change, the relayer would only be able to check the app_hash and/or the validator set, which might be sufficient for detection, but it wouldn't be able to verify the signed header directly. To get the signed header, it would have to find the corresponding UpdateClient tx in the blockchain (not the state) and extract the signed header (and potentially a bisection trace) from that. I'm not sure if this is straight forward right now, but something we'll have to enable either through events or otherwise.

Reply:
If I understand correctly, to convince the handler that there is a fork we can still submit the conflicting header. So this can stay the same.  
To convince a full node that there is a fork, we need to work on extracting the signed header.